File: kata-catalog/Supermarket_checkout.py
class Checkout:
    def __init__(self):
        self.prices = {
            'SOUP': {'unit_price': 1.00, 'volume_rule': {'quantity': 2, 'free_quantity': 1}},
            'RAMEN': {'unit_price': 0.40, 'volume_rule': {'quantity': 3, 'total_price': 1.00}},
            'ORANGE': {'unit_price': 2.00},
            'GRAPES': {'unit_price': 4.00, 'lbs': 0.00},       # no price_per_unit needed, scale used
            'YOGURT': {'unit_price': 0.67, 'volume_rule': {'quantity': 3, 'total_price': 2.00}}
        }
        self.cart = {}  

    def scan(self, sku, quantity=1):
        if sku not in self.prices:
            return 'Invalid item SKU'

        if sku == 'GRAPES':
            quantity = float(input('How many lbs?'))

        if sku not in self.cart:
            self.cart[sku] = {'quantity': 0, 'price': 0.00}
        
        unit_price = self.prices[sku].get('unit_price', None)
        volume_rule = self.prices[sku].get('volume_rule', None)

        self.cart[sku]['quantity'] += quantity
        self.cart[sku]['price'] += quantity*unit_price

        display_str = f"{sku} x {self.cart[sku]['quantity']}: "

        if volume_rule is not None:
            total_price = volume_rule.get('total_price', None)
            display_str += f"${unit_price:.2f} each"
            if self.cart[sku]['quantity'] % volume_rule['quantity'] == 0:
                if total_price is not None:
                    self.cart[sku]['price'] = total_price
                    display_str = f"{sku} x {self.cart[sku]['quantity']}: ${total_price:.2f} total"
                else:
                    self.cart[sku]['price'] -= unit_price
                    display_str = f"{sku} x {self.cart[sku]['quantity']}: $0.00 each"

        else:
            display_str += f"${self.cart[sku]['price']:.2f}"

        return display_str

    def total(self):
        total_price = sum(item['price'] for item in self.cart.values())
        return f"Total: ${total_price:.2f}"

File: kata-catalog/Supermarket_checkout.cs
using System;
using System.Collections.Generic;

public class Checkout
{
    public string sku;
    public float total;
    public static Dictionary<string, int> cart = new Dictionary<string, int>();

    public void Initialise_cart()
    {
        cart.Add("SOUP", 0);
        cart.Add("RAMEN", 0);
        cart.Add("ORANGE", 0);
        cart.Add("GRAPES", 0);
        cart.Add("YOGURT", 0);
    }

    public string tot()
    {
        return "Total: $" + total;
    }

    public string Scan(string sku)
    {
        int offer = 0;
        float unit_price = 0;
        string display_str = " ";
        switch (sku)
        {
            case "SOUP":
                cart[sku]++;
                offer = cart[sku] % 2;
                switch (offer)
                {
                    case 0:
                        unit_price = 0.00f;
                        break;
                    default:
                        unit_price = 1.00f;
                        break;
                }
                total += unit_price;
                display_str = sku + " x " + cart[sku] + ": $" + unit_price;
                return display_str;

            case "RAMEN":
                cart[sku]++;
                offer = cart[sku] % 3;
                switch (offer)
                {
                    case 0:
                        unit_price = 0.20f;
                        break;
                    default:
                        unit_price = 0.40f;
                        break;
                }
                total += unit_price;
                display_str = sku + " x " + cart[sku] + ": $" + unit_price;
                return display_str;
            
            case "ORANGE":
                cart[sku]++;
                unit_price = 2.00f;
                total += unit_price;
                display_str = sku + " x " + cart[sku] + ": $" + unit_price;
                return display_str;

            case "GRAPES":
                cart[sku]++;
                unit_price = 4.00f;
                total += unit_price;
                display_str = sku + " x " + cart[sku] + ": $" + unit_price;
                return display_str;

            case "YOGURT":
                cart[sku]++;
                offer = cart[sku] % 3;
                switch (offer)
                {
                    case 0:
                        unit_price = 0.66f;
                        break;
                    default:
                        unit_price = 0.67f;
                        break;
                }
                total += unit_price;
                display_str = sku + " x " + cart[sku] + ": $" + unit_price;
                return display_str;

            default:
                return "Invalid item SKU";
        }
    }
}
